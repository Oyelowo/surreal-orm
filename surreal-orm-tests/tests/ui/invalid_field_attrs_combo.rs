use serde::{Deserialize, Serialize};
use surreal_orm::{
    statements::{define_field, for_, select, DefineFieldStatement, Permissions},
    *,
};

// There should nne 12 errors in this file.
// And there were as at August 19, 2023. If you add more, please update this comment.
// The errors are from StudentTest0 to StudentTest11:
// 1. 1 Invalid value and value_fn combination
// 2. 1 Invalid assert and assert_fn combination
// 3. 1 Invalid permissions and permissions_fn combination
// 4. 1 Invalid define and define_fn combination
// 5. 8 Invalid combinations. Other attributes should not be used when define or define_fn is used
//    except for type

fn get_age_default_value() -> u8 {
    18
}

fn get_age_assertion() -> Filter {
    cond(value().is_not(NONE))
}

fn age_permissions() -> Permissions {
    for_([CrudType::Create, CrudType::Delete])
        .where_(StudentTest0::schema().firstName.is("Oyelowo"))
        .into()
}

fn student_permissions() -> Permissions {
    for_([CrudType::Create, CrudType::Delete])
        .where_(StudentTest0::schema().firstName.is("Oyelowo"))
        .into()
}

fn define_age() -> DefineFieldStatement {
    use surreal_orm::{Model, Node};
    use CrudType::*;
    let studenttest0_schema::StudentTest0 {
        ref age,
        ref firstName,
        ..
    } = StudentTest0::schema();

    use FieldType::*;

    let statement = define_field(age)
        .on_table(StudentTest0::table_name())
        .type_(Int)
        .value("oyelowo@codebreather.com")
        .assert(cond(value().is_not(NONE)).and(value().like("is_email")))
        .permissions(for_(Select).where_(age.greater_than_or_equal(18))) // Single works
        .permissions(for_([Create, Update]).where_(firstName.is("Oyedayo"))) //Multiple
        .permissions([
            for_([Create, Delete]).where_(firstName.is("Oyelowo")),
            for_(Update).where_(age.less_than_or_equal(130)),
        ]);
    statement
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_0")]
pub struct StudentTest0 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        value = "18",
        assert = "cond(value().is_not(NONE)).and(value().gte(18))",
        permissions = "for_([CrudType::Create, CrudType::Delete]).where_(StudentTest0::schema().firstName.is(\"Oyelowo\"))",
        define = "define_age()"
    )]
    age: u8,
    // #[surreal_orm(
    //     type = "int",
    //     value = "get_age_default_value()",
    //     assert = "get_age_assertion()",
    //     permissions = "age_permissions()",
    //     // define = "define_age()"
    // )]
    // age_default_external_function_invoked_expr: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_1")]
pub struct StudentTest1 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(type = "int", value = "18", define = "define_age()")]
    age: u8,
}

// do for others
#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_2")]
pub struct StudentTest2 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        assert = "cond(value().is_not(NONE)).and(value().gte(18))",
        define = "define_age()"
    )]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_3")]
pub struct StudentTest3 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        permissions = "for_([CrudType::Create, CrudType::Delete]).where_(StudentTest3::schema().firstName.is(\"Oyelowo\"))",
        define = "define_age()"
    )]
    age: u8,
}

// Do same with define_fn
#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_4")]
pub struct StudentTest4 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        value = "18",
        assert = "cond(value().is_not(NONE)).and(value().gte(18))",
        permissions = "for_([CrudType::Create, CrudType::Delete]).where_(StudentTest4::schema().firstName.is(\"Oyelowo\"))",
        define = "define_age()"
    )]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_5")]
pub struct StudentTest5 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(type = "int", value = "18", define_fn = "define_age")]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_6")]
pub struct StudentTest6 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        assert = "cond(value().is_not(NONE)).and(value().gte(18))",
        define_fn = "define_age"
    )]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_7")]
pub struct StudentTest7 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        permissions = "for_([CrudType::Create, CrudType::Delete]).where_(StudentTest7::schema().firstName.is(\"Oyelowo\"))",
        define_fn = "define_age"
    )]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_8")]
pub struct StudentTest8 {
    id: SurrealSimpleId<Self>,
    #[surreal_orm(
        type = "int",
        define = "define_age()"
        define_fn = "define_age"
    )]
    age: u8,
}

// normal attribute and its counterpart suffixed with _fn should not be allowed
#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_9")]
pub struct StudentTest9 {
    #[surreal_orm(
        type = "int",
        value = "get_age_default_value()",
        value_fn = "get_age_default_value"
    )]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_10")]
pub struct StudentTest10 {
    #[surreal_orm(
        type = "int",
        assert = "get_age_assertion()",
        assert_fn = "get_age_assertion"
    )]
    age: u8,
}

#[derive(Node, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student_test_11")]
pub struct StudentTest11 {
    #[surreal_orm(
        type = "int",
        permissions = "age_permissions()",
        permissions_fn = "age_permissions"
    )]
    age: u8,
}
