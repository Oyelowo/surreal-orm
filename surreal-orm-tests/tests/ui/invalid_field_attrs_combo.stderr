error: proc-macro derive panicked
  --> tests/ui/invalid_field_attrs_combo.rs:62:10
   |
62 | #[derive(Node, Serialize, Deserialize)]
   |          ^^^^
   |
   = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
               assert,
               assert_fn,
               value,
               value_fn,
               permissions,
               permissions_fn,
               item_assert,
               item_assert_fn

error: proc-macro derive panicked
  --> tests/ui/invalid_field_attrs_combo.rs:85:10
   |
85 | #[derive(Node, Serialize, Deserialize)]
   |          ^^^^
   |
   = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
               assert,
               assert_fn,
               value,
               value_fn,
               permissions,
               permissions_fn,
               item_assert,
               item_assert_fn

error: proc-macro derive panicked
  --> tests/ui/invalid_field_attrs_combo.rs:95:10
   |
95 | #[derive(Node, Serialize, Deserialize)]
   |          ^^^^
   |
   = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
               assert,
               assert_fn,
               value,
               value_fn,
               permissions,
               permissions_fn,
               item_assert,
               item_assert_fn

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:108:10
    |
108 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
                assert,
                assert_fn,
                value,
                value_fn,
                permissions,
                permissions_fn,
                item_assert,
                item_assert_fn

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:122:10
    |
122 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
                assert,
                assert_fn,
                value,
                value_fn,
                permissions,
                permissions_fn,
                item_assert,
                item_assert_fn

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:137:10
    |
137 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
                assert,
                assert_fn,
                value,
                value_fn,
                permissions,
                permissions_fn,
                item_assert,
                item_assert_fn

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:146:10
    |
146 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
                assert,
                assert_fn,
                value,
                value_fn,
                permissions,
                permissions_fn,
                item_assert,
                item_assert_fn

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:159:10
    |
159 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: Invalid combination. When `define` or `define_fn`, the following attributes cannot be use in combination to prevent confusion:
                assert,
                assert_fn,
                value,
                value_fn,
                permissions,
                permissions_fn,
                item_assert,
                item_assert_fn

error: expected `,`
   --> tests/ui/invalid_field_attrs_combo.rs:180:9
    |
180 |         define_fn = "define_age"
    |         ^^^^^^^^^

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:186:10
    |
186 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: value and value_fn attribute cannot be provided at the same time to prevent ambiguity. Use either of the two.

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:198:10
    |
198 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: assert and assert_fn attribute cannot be provided at the same time to prevent ambiguity. Use either of the two.

error: proc-macro derive panicked
   --> tests/ui/invalid_field_attrs_combo.rs:210:10
    |
210 | #[derive(Node, Serialize, Deserialize)]
    |          ^^^^
    |
    = help: message: permissions and permissions_fn attribute cannot be provided at the same time to prevent ambiguity. Use either of the two.

warning: unused import: `select`
 --> tests/ui/invalid_field_attrs_combo.rs:3:38
  |
3 |     statements::{define_field, for_, select, DefineFieldStatement, Permissions},
  |                                      ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `Node`
  --> tests/ui/invalid_field_attrs_combo.rs:38:30
   |
38 |     use surreal_orm::{Model, Node};
   |                              ^^^^

warning: the item `Model` is imported redundantly
  --> tests/ui/invalid_field_attrs_combo.rs:38:23
   |
4  |     *,
   |     - the item `Model` is already imported here
...
38 |     use surreal_orm::{Model, Node};
   |                       ^^^^^

warning: the item `Node` is imported redundantly
  --> tests/ui/invalid_field_attrs_combo.rs:38:30
   |
4  |     *,
   |     - the item `Node` is already imported here
...
38 |     use surreal_orm::{Model, Node};
   |                              ^^^^

error[E0601]: `main` function not found in crate `$CRATE`
   --> tests/ui/invalid_field_attrs_combo.rs:220:2
    |
220 | }
    |  ^ consider adding a `main` function to `$DIR/tests/ui/invalid_field_attrs_combo.rs`

error[E0277]: the trait bound `StudentTest0: surreal_orm::Model` is not satisfied
  --> tests/ui/invalid_field_attrs_combo.rs:66:9
   |
66 |     id: SurrealSimpleId<Self>,
   |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest0`
   |
   = help: the following other types implement trait `surreal_orm::Model`:
             TestUser
             surreal_models::Account
             surreal_models::Alien
             surreal_models::Alien2
             surreal_models::AlienWithExplicitAttributes
             surreal_models::Balance
             surreal_models::Blog
             surreal_models::Book
           and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
  --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
   |
   | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
   |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest1: surreal_orm::Model` is not satisfied
  --> tests/ui/invalid_field_attrs_combo.rs:89:9
   |
89 |     id: SurrealSimpleId<Self>,
   |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest1`
   |
   = help: the following other types implement trait `surreal_orm::Model`:
             TestUser
             surreal_models::Account
             surreal_models::Alien
             surreal_models::Alien2
             surreal_models::AlienWithExplicitAttributes
             surreal_models::Balance
             surreal_models::Blog
             surreal_models::Book
           and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
  --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
   |
   | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
   |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest2: surreal_orm::Model` is not satisfied
  --> tests/ui/invalid_field_attrs_combo.rs:99:9
   |
99 |     id: SurrealSimpleId<Self>,
   |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest2`
   |
   = help: the following other types implement trait `surreal_orm::Model`:
             TestUser
             surreal_models::Account
             surreal_models::Alien
             surreal_models::Alien2
             surreal_models::AlienWithExplicitAttributes
             surreal_models::Balance
             surreal_models::Blog
             surreal_models::Book
           and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
  --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
   |
   | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
   |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest3: surreal_orm::Model` is not satisfied
   --> tests/ui/invalid_field_attrs_combo.rs:112:9
    |
112 |     id: SurrealSimpleId<Self>,
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest3`
    |
    = help: the following other types implement trait `surreal_orm::Model`:
              TestUser
              surreal_models::Account
              surreal_models::Alien
              surreal_models::Alien2
              surreal_models::AlienWithExplicitAttributes
              surreal_models::Balance
              surreal_models::Blog
              surreal_models::Book
            and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
   --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
    |
    | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
    |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest4: surreal_orm::Model` is not satisfied
   --> tests/ui/invalid_field_attrs_combo.rs:126:9
    |
126 |     id: SurrealSimpleId<Self>,
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest4`
    |
    = help: the following other types implement trait `surreal_orm::Model`:
              TestUser
              surreal_models::Account
              surreal_models::Alien
              surreal_models::Alien2
              surreal_models::AlienWithExplicitAttributes
              surreal_models::Balance
              surreal_models::Blog
              surreal_models::Book
            and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
   --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
    |
    | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
    |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest5: surreal_orm::Model` is not satisfied
   --> tests/ui/invalid_field_attrs_combo.rs:141:9
    |
141 |     id: SurrealSimpleId<Self>,
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest5`
    |
    = help: the following other types implement trait `surreal_orm::Model`:
              TestUser
              surreal_models::Account
              surreal_models::Alien
              surreal_models::Alien2
              surreal_models::AlienWithExplicitAttributes
              surreal_models::Balance
              surreal_models::Blog
              surreal_models::Book
            and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
   --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
    |
    | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
    |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest6: surreal_orm::Model` is not satisfied
   --> tests/ui/invalid_field_attrs_combo.rs:150:9
    |
150 |     id: SurrealSimpleId<Self>,
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest6`
    |
    = help: the following other types implement trait `surreal_orm::Model`:
              TestUser
              surreal_models::Account
              surreal_models::Alien
              surreal_models::Alien2
              surreal_models::AlienWithExplicitAttributes
              surreal_models::Balance
              surreal_models::Blog
              surreal_models::Book
            and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
   --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
    |
    | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
    |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest7: surreal_orm::Model` is not satisfied
   --> tests/ui/invalid_field_attrs_combo.rs:163:9
    |
163 |     id: SurrealSimpleId<Self>,
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest7`
    |
    = help: the following other types implement trait `surreal_orm::Model`:
              TestUser
              surreal_models::Account
              surreal_models::Alien
              surreal_models::Alien2
              surreal_models::AlienWithExplicitAttributes
              surreal_models::Balance
              surreal_models::Blog
              surreal_models::Book
            and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
   --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
    |
    | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
    |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0277]: the trait bound `StudentTest8: surreal_orm::Model` is not satisfied
   --> tests/ui/invalid_field_attrs_combo.rs:176:9
    |
176 |     id: SurrealSimpleId<Self>,
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `surreal_orm::Model` is not implemented for `StudentTest8`
    |
    = help: the following other types implement trait `surreal_orm::Model`:
              TestUser
              surreal_models::Account
              surreal_models::Alien
              surreal_models::Alien2
              surreal_models::AlienWithExplicitAttributes
              surreal_models::Balance
              surreal_models::Blog
              surreal_models::Book
            and $N others
note: required by a bound in `surreal_orm::SurrealSimpleId`
   --> $WORKSPACE/surreal-query-builder/src/types/surreal_id.rs
    |
    | pub struct SurrealSimpleId<T: Model>(SurrealId<T, String>);
    |                               ^^^^^ required by this bound in `SurrealSimpleId`

error[E0433]: failed to resolve: use of undeclared crate or module `studenttest0_schema`
  --> tests/ui/invalid_field_attrs_combo.rs:40:9
   |
40 |     let studenttest0_schema::StudentTest0 {
   |         ^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `studenttest0_schema`
